R/caretStack.R

```diff
@@ -100,12 +123,90 @@ caretStack <- function(
     models = all.models,
     ens_model = model,
     error = model$results,
-    excluded_class_id = excluded_class_id
+    excluded_class_id = excluded_class_id,
+    original_features_included = which.var
   )
   class(out) <- "caretStack"
   out
 }
 
+#' @title Check the parameters use_original_features, new_X, and which.var are consistent
+#' @description Check the parameters use_original_features, new_X, and which.var are consistent
+#' when training a caretStack.
+#' @param use_original_features a logical indicating whether to include the original features in the stack.
+#' @param new_X a data.frame of new data to train the stack on.
+#' @param which.var a vector indicating which original features to include in the stack.
+#' @keywords internal
+check_use_original_features <- function(use_original_features, new_X, which.var) {
+  # Make sure new_X is not NULL when use_original_features is not NULL
+  if (use_original_features && is.null(new_X)) {
+    warning("use_original_features is ignored when new_X is NULL.", call. = FALSE)
+    use_original_features <- FALSE
+    which.var <- NULL
+  }
+
+  # Advertise when which.var is ignored
+  if (!use_original_features && !is.null(which.var)) {
+    warning("which.var is ignored when use_original_features is FALSE.", call. = FALSE)
+    which.var <- NULL
+  }
+
+  list(use_original_features = use_original_features, which.var = which.var)
+}
+
+#' @title Process which.var
+#' @description Process the which.var argument for caretStack.
+#' @param use_original_features a logical indicating whether to include the original features in the stack.
+#' @param feature_colnames a character vector of the names of the original features used to train the stack.
+#' @param which.var a character vector of the names of the original features to include in the stack or a numeric
+#' vector with the indexes of the original features to include in the stack. If NULL, will include all original
+#' features. If use_original_features is FALSE, this argument is ignored.
+#' @keywords internal
+process_which_var <- function(use_original_features, feature_colnames, which.var) {
+  # If use_original_features are not used, which.var is ignored
+  if (!use_original_features) {
+    return(NULL)
+  }
+  # Use all original features when which.var is NULL
+  if (is.null(which.var)) {
+    return(feature_colnames)
+  }
+
+  if (is.character(which.var)) {
+    # Check that all specified variables in which.var are colnames of
+    ignored_vars <- setdiff(which.var, feature_colnames)
+    if (length(ignored_vars) > 0L) {
+      warning("The following variables are not in new_X and will be ignored: ",
+        toString(ignored_vars), ". Ignoring these variables.",
+        call. = FALSE
+      )
+      return(setdiff(which.var, ignored_vars))
+    }
+    return(which.var)
+  }
+  if (is.numeric(which.var)) {
```

Comment:
lets require the features specified to be a character vector, rather than maintaining 2 sets of logic

Reply:
Ok